[Mobile] Add additional security to wallet key storage

cmcewen opened the pull request:
### Description

To improve security of key storage, this generates a unique client pepper and store in the keystore/keychain to provide more entropy to the blockchain. It, combined with a user pin, is used to encrypt the private key on disk. This prevents an attacker with disk access from easily getting the private key

_[Rossy edit]:_

Also blacklisted unsafe pins, sanitized logs, fixed issue with pin caching while navigating, renamed lockWithPinEnabled to requirePinOnAppOpen, and some refactoring.

### Other changes

Store the mnemonic encrypted in the keystore.

### Tested

Run locally on android and iOS

_[Rossy edit]:_

Tested with adding, removing, and changing PIN settings on Android, confirmed the keystore values were still accessible. Mnemonic decrypts correctly and account can be unlocked.

Tested changing PIN on iOS, confirmed the same. Tested with no PIN on iOS with simulator, all good.

cmcewen committed:
WIP

cmcewen committed:
wip

cmcewen committed:
wip

cmcewen committed:
Merge branch 'master' into cmcewen/keystore

cmcewen committed:
random bytes to string

cmcewen committed:
Merge branch 'master' into cmcewen/keystore

cmcewen committed:
Properly store and fetch hash

jmrossy commented on packages/mobile/__mocks__/react-native-keychain.ts:
```diff
@@ -0,0 +1,10 @@
+  resetGenericPassword: jest.fn(),
+}
+
+export default keychainMocka
```
typo?

jmrossy commented on packages/mobile/android/app/.settings/org.eclipse.buildship.core.prefs:
```diff
@@ -1,2 +1,13 @@
-connection.project.dir=..
+arguments=
```
Does this file just affect eclipse users?

jmrossy commented on packages/mobile/android/build.gradle:
```diff
@@ -4,8 +4,8 @@ buildscript {
         minSdkVersion = 16
-        compileSdkVersion = 28
-        targetSdkVersion = 28
+        compileSdkVersion = 29
```
why bump to 29?

cmcewen replied:
This was necessary for react-native-keychain to work properly

jmrossy commented on packages/mobile/src/account/saga.ts:
```diff
@@ -14,25 +13,13 @@ import { CustomEventNames } from 'src/analytics/constants'
-      Logger.debug(TAG + '@setPincode', 'Setting pincode with using system auth')
-      pin = randomBytes(10).toString('hex') as string
-      yield call(setPinInKeystore, pin)
-    } else if (pincodeType === PincodeType.CustomPin && pin) {
```
with this gone, what's this function supposed to be doing now?

cmcewen replied:
I left it so that we could easily add biometrics in later

jmrossy commented on packages/mobile/src/pincode/PincodeSet.tsx:
```diff
@@ -69,12 +69,18 @@ export class PincodeSet extends React.Component<Props, State> {
+        this.props.setPincode(PincodeType.CustomPin)
+        this.props.navigation.navigate(Screens.EnterInviteCode)
+      } catch (err) {
+        CeloAnalytics.track(CustomEventNames.pin_store_error, { error: err.message })
```
Might wanna confirm the pin's not in this message

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,99 @@
+export const PIN_LENGTH = 6
+
+let cachedPepper: string
+const cachedPasswordHashes: {
```
Wondering if this should go with the other cached stuff in your new passwordCache file?

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,99 @@
+}
+
+async function hashForPin(pin: string) {
+  return sha256(new Buffer(await passwordForPin(pin), 'hex')).toString('hex')
```
Should we use a multi-iteration hash here for more security?

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,99 @@
+  if (hash === passwordHash) {
+    return true
+  }
+  return false
```
```suggestion
  return hash === passwordHash
```

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,99 @@
+    }
+    const unlocked = await ensureCorrectPin(await passwordForPin(pin), currentAccount)
+    if (unlocked) {
+      await storePasswordHash(pin, account)
```
Why do we store the password hash instead of recomputing it with the pin each time?

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,99 @@
+  if (cachedPepper) {
+    return cachedPepper
+  }
+  const storedPepper = await retrieveStoredItem(STORAGE_KEYS.PEPPER)
```
Should we have error handling at some layer or is it intentional to allow them to propagate up to the component?

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,99 @@
+  const passwordHash = await retrievePasswordHash(account)
+  if (!passwordHash) {
+    Logger.error(`${TAG}@checkPin`, 'No password hash stored')
+    const currentAccount = currentAccountSelector(store.getState())
```
Any reason this new code isn't in a saga? Is it just to avoid the boilerplate?

cmcewen replied:
I just want to avoid any password handling getting accidentally logged, so trying to keep it as contained as possible

jmrossy commented on packages/mobile/src/utils/keyStore.ts:
```diff
@@ -8,8 +8,9 @@ export async function setKey(key: string, value: string) {
     // TODO(Rossy + Jean): Revisit this accessible setting
-    await RNSecureKeyStore.set(key, value, {
-      accessible: ACCESSIBLE.WHEN_UNLOCKED,
+    await Keychain.setGenericPassword('CELO', value, {
```
We should probably folder this code into your new storage/keychain file right?

jmrossy commented on packages/mobile/src/web3/saga.ts:
```diff
@@ -259,12 +263,14 @@ export function* unlockAccount(account: string) {
-    const pincode = yield call(getPincode, true)
-    yield call([wallet, wallet.unlockAccount], account, pincode, UNLOCK_DURATION)
+    const pin = yield call(getPincode, true)
+    const password: string = yield call(passwordForPin, pin)
```
These two seem to be always called together. Should they be merged or wrapped?

cmcewen committed:
Merge branch 'master' into cmcewen/keystore

cmcewen committed:
Address PR comments

cmcewen committed:
Remove unused .settings

cmcewen committed:
Clean up

jeanregisser commented on packages/mobile/src/backup/utils.ts:
```diff
@@ -162,3 +170,14 @@ export function isValidBackupPhrase(phrase: string) {
+
+export async function encryptMnemonic(phrase: string) {
+  const pepper = await retrieveOrGeneratePepper()
+  return CryptoJS.AES.encrypt(phrase, pepper).toString()
```
Shouldn't we encrypt the mnemonic with both the pepper and passwordHash?

jmrossy replied:
I'll be changing this to use the password instead, which is pin + pepepr

jeanregisser reviewed (approved):
Awesome work and effort for tightening the security of the wallet! 🥇 🎉  
Left one question.

jmrossy commented on packages/mobile/src/backup/utils.ts:
```diff
@@ -162,3 +170,14 @@ export function isValidBackupPhrase(phrase: string) {
+
+export async function encryptMnemonic(phrase: string) {
+  const pepper = await retrieveOrGeneratePepper()
+  return CryptoJS.AES.encrypt(phrase, pepper).toString()
```
Do we need to specify the encoding here too like line 182?

jmrossy commented on packages/mobile/src/backup/utils.ts:
```diff
@@ -162,3 +170,14 @@ export function isValidBackupPhrase(phrase: string) {
 }
+
+export async function encryptMnemonic(phrase: string) {
+  const pepper = await retrieveOrGeneratePepper()
```
I thought we were going to use the pin to encrypt the mnemonic in case the pepper is lost before the user does the backup flow?

jmrossy replied:
Using password instead.

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,151 @@
+  return pin.length === PIN_LENGTH
+}
+
+function pinStorageKey(account: string) {
```
The name of this confused me. It's actually the account password hash storage key right?

jmrossy commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -0,0 +1,151 @@
+  return getCachedPasswordHash(account)
+}
+
+type PinPromise = (pin: string) => void
```
PinCallback?

jmrossy committed:
Merge branch 'master' into cmcewen/keystore

jmrossy committed:
Complete implementation of new pin + password scheme
Reorganize new authentication code
Add timeouts to all cached secrets
Blacklist obvious pins
Rename lockWithPinEnabled to requirePinOnAppOpen
Reduce unlock default account unlock duration from 10 to 5 minutes
Sanitize log messages that may contain secrets
Fix tests and add new saga test for getting password

jmrossy commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -0,0 +1,86 @@
+})
+
+describe(getPincode, () => {
+  // it('returns PIN from phone keystore', async () => {
```
Copied over from the old `getPincode` tests in the now deleted account/saga.test file

tarikbellamine commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -1,94 +1,109 @@
+// Using this value in the caches
+export const DEFAULT_CACHE_ACCOUNT = 'default'
+
+const PIN_BLACKLIST = [
```
May be worth creating/throwing a custom error message for this to surface to the user. IMO may not be obvious why these are invalid.

Also would be nice to disable this with a dev flag because we've just blacklisted all of my go-tos ;)

jmrossy replied:
Yeah that's fair but I'm going to punt that to a follow-up PR. Juggling a few things atm and I think it's important to land this asap cause its risky.

jmrossy replied:
For the moment it shows 'Invalid Pin' to the user

jmrossy committed:
Merge branch 'master' into cmcewen/keystore

jmrossy committed:
Fix snapshot test

cmcewen committed:
Merge branch 'master' into cmcewen/keystore

cmcewen committed:
Commit podfile

i1skn commented on packages/react-components/services/ReactNativeLogger.ts:
```diff
@@ -58,6 +58,20 @@ export default class ReactNativeLogger {
+    }
+
+    if (valueToPurge) {
+      return new Error(message.replace(new RegExp(valueToPurge, 'g'), ''))
```
Do we want to place something like `<purged>` instead of `''` here?

jmrossy replied:
Will do

i1skn reviewed (approved):
Great work!! Only one little comment.

jmrossy committed:
Merge branch 'master' into cmcewen/keystore

jmrossy committed:
Encrypt mnemonic with password instead of pin
Fix bugs
Improve sanitize error interface

jmrossy committed:
Fix ios e2e test

jmrossy committed:
Merge branch 'master' into cmcewen/keystore

jmrossy committed:
Merge branch 'master' into cmcewen/keystore

jmrossy committed:
Merge branch 'master' into cmcewen/keystore
